### ISSUE TITLE
pyright failed to create stub for scipy.

### ISSUE THREAD

**Issue by pittcat** [microsoft/pyright#2324]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright does not create a full stub for scipy, only a partial one.

**To Reproduce**
Use `pyright --createstub scipy` from the command line or coc-pyright, and you will see that this process never ends.

**Expected behavior**
Scipy stub can be create by pyright.

**Comment by erictraut** [comment]
Scipy is untyped and has code that is extremely complex and costly to analyze. It should eventually complete, but it could take a very long time. Because of the approach that pyright uses for analysis, there's probably not much that can be done to improve this.

I recommend that you use the partial stubs from [here](https://github.com/microsoft/python-type-stubs/tree/main/scipy-stubs). These are the stubs we ship with pylance for scipy.

**Comment by erictraut** [comment]
@pittcat, do those stubs look like they'll meet your needs?

I'm going to keep this issue open and take another look at the perf issues with scipy to see if there's something more that can be done.

**Comment by pittcat** [comment]
I have tried python-type-stubs, but It doesn't seem to work satisfactorily. I would still choose to use copyright to generate the stub, even if it might take a little longer.

**Comment by erictraut** [comment]
There were two issues contributing to the long times when generating scipy type stubs:
1. There was a performance bug in the type stub generation code where it was unnecessarily evaluating the types of local variables, which don't get emitted as part of the stubs anyway
2. Some functions in scipy are very complex, and because they have no type annotations (even for input parameters), the type of every symbol must be inferred. This leads to very long analysis times when inferring return types.

I've fixed the bug for the first issue. For the second issue, I've added a heuristic to skip return type inference for extremely complex functions. With these two fixes in place, I'm now able to generate type stubs for scipy.

These fixes will be in the next release of pyright.
